lib/concurrent/synchronization/java_pure_object.rb

```diff
@@ -32,6 +33,10 @@ def ns_signal
           JRuby.reference0(self).notify
           self
         end
+
+        def ensure_ivar_visibility!
+          # relying on undocumented behavior of JRuby, ivar access is volatile
```

Comment:
I had a conversation on FOSDEM with @headius about the instance variables and it turned out that the access is in fact volatile, by default it uses https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/StampedVariableAccessor.java even though the documentation says that they are not. I got the feeling that the reason is to be able to go back to non-volatile ivars if needed. There are also experimental options like `REIFY_VARIABLES` which uses https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/VariableAccessorVar0.java where the access is not volatile. (I should probably check for it in the synchronization layer). Advantage of having this layer is that we can react to the changes in the implementations still providing unified behavior, without making changes in the rest of the `concurrent` library.

Just for completeness: ivars are volatile on MRI and JRuby but not on Rubinius. (AFAIK)

Exactly. I am already planning to finish this layer, write down documentation (including the assumptions it builds on) and then contact the implementors to be able to verify the assumptions and therefore soundness of this layer.